File: SalesTaxes/Controller/ShoppingViewController.m
//
//  ShoppingCartViewController.m
//  SalesTaxes
//
//  Created by Fabrizio Duroni on 11/11/2016.
//  Copyright © 2016 Fabrizio Duroni. All rights reserved.
//

#import "ShoppingViewController.h"
#import "ReceiptViewController.h"
#import "ShoppingCell.h"
#import "Product.h"
#import "ShoppingItem.h"
#import "ShoppingCart.h"
#import "Receipt.h"

@interface ShoppingViewController ()

/// Product list that could be selected.
@property (nonatomic, strong) NSArray<Product *> *productList;
/// Shopping cart.
@property (nonatomic, strong) ShoppingCart *shoppingCart;

@end

@implementation ShoppingViewController

- (void)viewDidLoad {
    
    [super viewDidLoad];
    
    //Static produc list (demo...).
    self.productList = @[
            [[Product alloc] initWithName:@"book" andPrice:12.49f andType:Book andOrigin:Local],
            [[Product alloc] initWithName:@"music" andPrice:14.99f andType:Other andOrigin:Local],
            [[Product alloc] initWithName:@"chocolate" andPrice:0.85f andType:Food andOrigin:Local],
            [[Product alloc] initWithName:@"imported box of chocolate"  andPrice:10.00f andType:Food andOrigin:Imported],
            [[Product alloc] initWithName:@"imported bottle of perfume chanel" andPrice:47.50f andType:Other andOrigin:Imported],
            [[Product alloc] initWithName:@"imported bottle of perfume armani" andPrice:27.99f andType:Other andOrigin:Imported],
            [[Product alloc] initWithName:@"bottle of perfume" andPrice:18.99f andType:Other andOrigin:Local],
            [[Product alloc] initWithName:@"packet of headachepills" andPrice:9.75f  andType:Medical andOrigin:Local],
            [[Product alloc] initWithName:@"box of imported chocolates" andPrice:11.25f andType:Food andOrigin:Imported]
    ];
    
    //Start shopping cart.
    self.shoppingCart = [[ShoppingCart alloc] init];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    
    return self.productList.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    ShoppingCell *cell = [tableView dequeueReusableCellWithIdentifier:@"ShoppingCell" forIndexPath:indexPath];
    
    Product *product = [self.productList objectAtIndex:indexPath.row];
    
    //Setup cell.
    cell.productName.text = product.productName;
    cell.productType.text = ProducTypeString(product.productType);
    cell.productOrigin.text = ProductOriginString(product.productOrigin);
    cell.productPrice.text = [NSString stringWithFormat:@"%.2f", product.productPrice];
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    UITableViewCell *cell = [tableView cellForRowAtIndexPath:indexPath];

    Product *product = [self.productList objectAtIndex:indexPath.row];

    if (cell.accessoryType == UITableViewCellAccessoryCheckmark) {

        //Delete.
        [self.shoppingCart remove:[product productName]];

        cell.accessoryType = UITableViewCellAccessoryNone;
    } else {
        
        UIAlertController * quantityAlertController = [UIAlertController alertControllerWithTitle: @"Quantity"
                                                                                          message: @"Insert quantity"
                                                                                   preferredStyle:UIAlertControllerStyleAlert];
        
        [quantityAlertController addTextFieldWithConfigurationHandler:^(UITextField *textField) {
            
            textField.placeholder = @"quantity";
            textField.text = @"1";
        }];
        
        [quantityAlertController addAction:[UIAlertAction actionWithTitle:@"OK"
                                                                    style:UIAlertActionStyleDefault
                                                                  handler:^(UIAlertAction *action) {
                                                                      
            NSArray * textfields = quantityAlertController.textFields;
            UITextField *quantityField = textfields[0];
            int quantity = [quantityField.text intValue];
                                                                      
            //If quantity is not valid insert default = 1.
            if (quantity == 0) {
                
                quantity = 1;
            }
                                                                      
            //Add.
            TaxCalculator *taxCalculator = [[TaxCalculator alloc] init];
            ShoppingItem *item = [[ShoppingItem alloc] initWithProduct:product
                                                           andQuantity:quantity
                                                      andTaxCalculator:taxCalculator];
                                                      
            [self.shoppingCart add:item];
                                                              
        }]];
        
        [self presentViewController:quantityAlertController animated:YES completion:nil];
        
        cell.accessoryType = UITableViewCellAccessoryCheckmark;
    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
 
    return 75.f;
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    
    if ([[segue identifier] isEqualToString:@"ReceiptSegue"]) {
        
        //Get reference to the destination view controller
        ReceiptViewController *receiptController = (ReceiptViewController *)segue.destinationViewController;
        receiptController.receipt = [[Receipt alloc] initWithShoppingCart:self.shoppingCart];
    }
}

- (void)didReceiveMemoryWarning {

    [super didReceiveMemoryWarning];
}


@end

File: SalesTaxesSwift/Controller/ShoppingViewController.swift
//
//  ViewController.swift
//  SalesTaxesSwift
//
//  Created by Fabrizio Duroni on 20/11/2016.
//  Copyright © 2016 Fabrizio Duroni. All rights reserved.
//

import UIKit

class ShoppingViewController: UITableViewController {
    
    /// Product list that could be selected.
    private var productList:[Product] = [Product]()
    /// Shopping cart.
    private var shoppingCart:ShoppingCart = ShoppingCart()
    
    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        //Static load (demo...)
        self.productList.append(Product(aName:"book", aPrice:12.49, aType:.Book, aOrigin:.Local))
        self.productList.append(Product(aName:"music", aPrice:14.99, aType:.Other, aOrigin:.Local))
        self.productList.append(Product(aName:"chocolate", aPrice:0.85, aType:.Food, aOrigin:.Local))
        self.productList.append(Product(aName:"imported box of chocolate", aPrice:10.00, aType:.Food, aOrigin:.Imported))
        self.productList.append(Product(aName:"imported perfume chanel", aPrice:47.50, aType:.Other, aOrigin:.Imported))
        self.productList.append(Product(aName:"imported perfume armani", aPrice:27.99, aType:.Other, aOrigin:.Imported))
        self.productList.append(Product(aName:"bottle of perfume", aPrice:18.99, aType:.Other, aOrigin:.Local))
        self.productList.append(Product(aName:"packet of headachepills", aPrice:9.75, aType:.Medical, aOrigin:.Local))
        self.productList.append(Product(aName:"box of imported chocolates", aPrice:11.25, aType:.Food, aOrigin:.Imported))
    }
    
    //MARK: UITableView delegate
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        
        return self.productList.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let cell:ShoppingCell = tableView.dequeueReusableCell(withIdentifier:"ShoppingCell", for:indexPath) as! ShoppingCell
        
        let product:Product = self.productList[indexPath.row]
        
        cell.productName.text = product.productName
        cell.productType.text = product.productType.rawValue
        cell.productOrigin.text = product.productOrigin.rawValue
        cell.productPrice.text = String(format:"%.2f", product.productPrice)
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        let cell:UITableViewCell = tableView.cellForRow(at: indexPath)!
        
        let product:Product = self.productList[indexPath.row]
        
        if cell.accessoryType == .checkmark {
            
            //Remove product from shopping cart.
            self.shoppingCart.remove(shoppingItemName: product.productName)
            
            cell.accessoryType = .none
        } else {
            
            //Add product to shopping cart.
            //Use UIAlertController to get the quantity for the product selected.
            let quantityAlertController = UIAlertController(title: "Quantity",
                                                            message: "Insert quantity",
                                                            preferredStyle: .alert)
            
            quantityAlertController.addTextField(configurationHandler:{ (textField) in
                
                textField.placeholder = "Quantity"
                textField.text = "1"
            })
            
            quantityAlertController.addAction(UIAlertAction(title:"OK",
                                                            style:UIAlertActionStyle.default,
                                                            handler:{ (action) in
                                                                
                    if let quantityField = quantityAlertController.textFields?[0] {
                                                                    
                        var quantity:Int = Int(quantityField.text!)!
                                                                    
                        if quantity == 0 {
                                                                        
                            quantity = 1
                        }
                                                                    
                        let item = ShoppingItem(aProduct: product, aQuantity:quantity, aTaxCalculator: TaxCalculator())
                        self.shoppingCart.add(aShoppingItem: item)
                    }
                })
            )
            
            self.present(quantityAlertController, animated: true, completion: nil)
            
            cell.accessoryType = .checkmark
        }
    }
    
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        
        return 75.0
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        
        if segue.identifier == "ReceiptSegue" {
            
            let receiptController:ReceiptViewController = segue.destination as! ReceiptViewController
            receiptController.receipt = Receipt(aShoppingCart: self.shoppingCart)
        }
    }
    
    override func didReceiveMemoryWarning() {
        
        super.didReceiveMemoryWarning()
    }
}
